### PR TITLE
[KYUUBI #3594][Subtask][Authz] init RangerBasePlugin for each catalog with corresponding service name

### LINKED ISSUE #3594: [Subtask] introduce Ranger Plugin factory and init RangerPlugin with service name for each catalog name
### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### Search before asking

- [X] I have searched in the [issues](https://github.com/apache/incubator-kyuubi/issues) and found no similar issues.

### Describe the subtask

introduce Ranger Plugin factory and init RangerPlugin with service name for each catalog name.

### Parent issue

#3586 

### Are you willing to submit PR?

- [X] Yes. I can submit a PR independently to improve.
- [ ] Yes. I would be willing to submit a PR with guidance from the Kyuubi community to improve.
- [ ] No. I cannot submit a PR at this time.

### PR DESCRIPTION
**Pull request by bowenliang123**
### _Why are the changes needed?_

to close #3594 .

- fetching Ranger service name by looking up catalog name in Ranger config of 'ranger.plugin.spark.catalog.$catalogName.service.name'
- init plugin with service name by catalog name
- existed usage of Authz plugin will not be impacted, as default value for ranger plugin's serviceType('spark') and appId('sparkSql') is not changed.

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by bowenliang123** [comment]
It shall pass the ut for Ranger 2.1+. The challenge is to adapt for Ranger 2.0 and below. 
- RangerBasePlugin constructor do not accept serviceName param
- RangerBasePlugin doesn't have `getConfig` method and no `RangerConfiguration` class

The first one is a major obstacle.

**Comment by bowenliang123** [comment]
for Ranger 2.1+, use constructor with service name , but for ranger 2.0 and below reuse same service name as default.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala [hunk 1]
```diff
@@ -131,4 +144,39 @@ object SparkRangerAdminPlugin extends RangerBasePlugin("spark", "sparkSql")
       }
     }
   }
+
+  def getRangerPlugin(catalog: Option[String] = None): RangerBasePlugin = {
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:148]: def getOrCreate()

Comment (yaooqinn) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:148]: shall we synchronize it?

Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:148]: ok, let's synchronize it.
Plus,  `catalog2pluginMap` itself is already using `TrieMap` as implementation , which is a concurrent threadsafe lock-free concurrent.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala [hunk 2]
```diff
@@ -131,4 +140,39 @@ object SparkRangerAdminPlugin extends RangerBasePlugin("spark", "sparkSql")
...
+      case None | Some("spark_catalog") =>
+        defaultBasePlugin
+      case Some(catalogName) =>
+        val serviceName = getRangerConf(defaultBasePlugin)
+          .get(s"ranger.plugin.spark.catalog.$catalogName.service.name")
+        serviceName match {
+          case _ if StringUtils.isBlank(serviceName) =>
+            logWarning(s"config ranger.plugin.spark.catalog.$catalogName.service.name not found," +
+              s" default ranger plugin is used")
+            defaultBasePlugin
```
Comment (bowenliang123) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:155]: Here returns a `defaultBasePlugin` for the catalog without setting the Ranger service name.
But it may shadow and block the normal access to the catalog.
How to return a allow-all ranger plugin to skip access checking for them?

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleApplyRowFilterAndDataMasking.scala [hunk 3]
```diff
@@ -81,17 +81,18 @@ class RuleApplyRowFilterAndDataMasking(spark: SparkSession) extends Rule[Logical
       plan: LogicalPlan,
       identifier: TableIdentifier,
       spark: SparkSession): LogicalPlan = {
+    val plugin = SparkRangerAdminPlugin
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleApplyRowFilterAndDataMasking.scala:84]: unnecessary？

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleAuthorization.scala [hunk 4]
```diff
@@ -87,11 +86,12 @@ object RuleAuthorization {
       }
     }
 
-    if (authorizeInSingleCall) {
-      verify(requestArrays.flatten, auditHandler)
+    val rangerPlugin = SparkRangerAdminPlugin
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleAuthorization.scala:89]: ditto

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala [hunk 5]
```diff
@@ -131,4 +140,39 @@ object SparkRangerAdminPlugin extends RangerBasePlugin("spark", "sparkSql")
       }
     }
   }
+
+  def getOrCreateRangerPlugin(catalog: Option[String] = None): RangerBasePlugin = synchronized {
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:144]: getOrCreate

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala [hunk 6]
```diff
@@ -38,12 +47,12 @@ object SparkRangerAdminPlugin extends RangerBasePlugin("spark", "sparkSql")
    * This configuration controls whether to verify the privilege objects in single call or
    * to verify them one by one.
    */
-  def authorizeInSingleCall: Boolean = getRangerConf.getBoolean(
-    s"ranger.plugin.${getServiceType}.authorize.in.single.call",
+  def authorizeInSingleCall: Boolean = getRangerConf(defaultBasePlugin).getBoolean(
+    s"ranger.plugin.${defaultBasePlugin.getServiceType}.authorize.in.single.call",
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:51]: why use the defaultBasePlugin.getServiceType?

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala (hunk 5 above)
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:144]: anywhere we set catalog?

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala [hunk 7]
```diff
@@ -131,4 +140,39 @@ object SparkRangerAdminPlugin extends RangerBasePlugin("spark", "sparkSql")
...
+    }
+  }
+
+  def initRangerBasePlugin(serviceName: String, appId: String): RangerBasePlugin = {
+    val basePlugin =
+      if (isRanger21orGreater) {
+        classOf[RangerBasePlugin].getConstructor(classOf[String], classOf[String], classOf[String])
+          .newInstance(serviceType, serviceName, appId)
+      } else {
+        // ignoring serviceName for Ranger 2.0 and below,
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:170]: Explain why not what

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/FilteredShowObjectsExec.scala [hunk 8]
```diff
@@ -47,7 +48,7 @@ case class FilteredShowNamespaceExec(delegated: SparkPlan) extends FilteredShowO
     val database = r.getString(0)
     val resource = AccessResource(ObjectType.DATABASE, database, null, null)
     val request = AccessRequest(resource, ugi, OperationType.SHOWDATABASES, AccessType.USE)
-    val result = SparkRangerAdminPlugin.isAccessAllowed(request)
+    val result = getOrCreateRangerPlugin().isAccessAllowed(request)
```
Comment (yaooqinn) [inline extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/FilteredShowObjectsExec.scala:51]: which service stores privileges for all catalogs if we want a filtered list of catalogs

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleApplyRowFilterAndDataMasking.scala (hunk 3 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/RuleApplyRowFilterAndDataMasking.scala:84]: Not really. Just to make sure `getFilterExpr` and `getMaskingExpr` are calling to same SparkRangerAdminPlugin instance. And separate them by catalog in future.

Or let me revert it  this time?

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala (hunk 5 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:144]: Alright. I put RangerPlugin as suffix here is to distinguish `SparkRangerAdminPlugin` itself and `RangerBasePlugin`.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala (hunk 6 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:51]: Compatible to exsiting config name as `ranger.plugin.spark.authorize.in.single.call`.
Or let me use `serviceType` with `spark` value instead.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala (hunk 7 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala:175]: Since it's not supporting different service names with same service type in Ranger 2.0 and below.
The only way to separate plugin config with service type is to seperate them in `ranger-security-${serviceType}.xml` configs file with different service types, plus repeating all the detail configs in it (e.g. REST url, user, password, and etc.). This is too heavy and hard to use.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala (hunk 5 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala]: But in usages, it will become `getOrCreate().isAccessAllowed(request)` which is misleading in some way.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/FilteredShowObjectsExec.scala (hunk 8 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/FilteredShowObjectsExec.scala]: No, this PR is not evolved for storing or checking users' privileges of catalog-level resources.
As for this line of the method, no action changed and it's filtering the database-level resources as in Hive service def, as the namespace-level in Spark concept.

File: extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala (hunk 5 above)
Comment (bowenliang123) [reply extensions/spark/kyuubi-spark-authz/src/main/scala/org/apache/kyuubi/plugin/spark/authz/ranger/SparkRangerAdminPlugin.scala]: This will be done in the following PR in pull <https://github.com/apache/incubator-kyuubi/pull/3606> by adding `catalog` to `PrivilegeObject`. And then fill catalog info from extractors in another PR.

The initial plan and steps are in the umberalla issue here . https://github.com/apache/incubator-kyuubi/issues/3586

**Comment by bowenliang123** [comment]
Also suspending this PR due to the unplanned close in pull https://github.com/apache/incubator-kyuubi/pull/3606 .
